tlv/primitive_test.go

```diff
@@ -0,0 +1,242 @@
+package tlv_test
+
+import (
+	"bytes"
+	"reflect"
+	"testing"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/lightningnetwork/lnd/tlv"
+)
+
+var testPK, _ = btcec.ParsePubKey([]byte{0x02,
+	0x3d, 0xa0, 0x92, 0xf6, 0x98, 0x0e, 0x58, 0xd2,
+	0xc0, 0x37, 0x17, 0x31, 0x80, 0xe9, 0xa4, 0x65,
+	0x47, 0x60, 0x26, 0xee, 0x50, 0xf9, 0x66, 0x95,
+	0x96, 0x3e, 0x8e, 0xfe, 0x43, 0x6f, 0x54, 0xeb,
+}, btcec.S256())
+
+type primitive struct {
+	u8    byte
+	u16   uint16
+	u32   uint32
+	u64   uint64
+	b32   [32]byte
+	b33   [33]byte
+	b64   [64]byte
+	pk    *btcec.PublicKey
+	bytes []byte
+}
+
+// TestWrongEncodingType asserts that all primitives encoders will fail with a
+// ErrTypeForEncoding when an incorrect type is provided.
+func TestWrongEncodingType(t *testing.T) {
+	encoders := []tlv.Encoder{
+		tlv.EUint8,
+		tlv.EUint16,
+		tlv.EUint32,
+		tlv.EUint64,
+		tlv.EBytes32,
+		tlv.EBytes33,
+		tlv.EBytes64,
+		tlv.EPubKey,
+		tlv.EVarBytes,
+	}
+
+	// We'll use an int32 since it is not a primitive type, which should
+	// cause the primitive encoders to fail with an ErrTypeForEncoding
+	// failure.
+	var (
+		value int32
+		buf   [8]byte
+		b     bytes.Buffer
+	)
+	for _, encoder := range encoders {
+		err := encoder(&b, &value, &buf)
+		if _, ok := err.(tlv.ErrTypeForEncoding); !ok {
+			t.Fatalf("expected error of type ErrTypeForEncoding, "+
+				"got %T", err)
+		}
+	}
+}
+
+// TestWrongDecodingType asserts that all primitives decoders will fail with a
+// ErrTypeForDecoding when an incorrect type is provided.
+func TestWrongDecodingType(t *testing.T) {
+	decoders := []tlv.Decoder{
+		tlv.DUint8,
+		tlv.DUint16,
+		tlv.DUint32,
+		tlv.DUint64,
+		tlv.DBytes32,
+		tlv.DBytes33,
+		tlv.DBytes64,
+		tlv.DPubKey,
+		tlv.DVarBytes,
+	}
+
+	// We'll use an int32 since it is not a primitive type, which should
+	// cause the primitive decoders to fail with an ErrTypeForDecoding
+	// failure.
+	var (
+		value int32
+		buf   [8]byte
+		b     bytes.Buffer
+	)
+	for _, decoder := range decoders {
+		err := decoder(&b, &value, &buf, 0)
+		if _, ok := err.(tlv.ErrTypeForDecoding); !ok {
+			t.Fatalf("expected error of type ErrTypeForDecoding, "+
+				"got %T", err)
+		}
+	}
+}
+
+type fieldEncoder struct {
+	val     interface{}
+	encoder tlv.Encoder
+}
+
+type fieldDecoder struct {
+	val     interface{}
+	decoder tlv.Decoder
+	size    uint64
+}
+
+// TestPrimitiveEncodings tests that we are able to serialize all known
+// primitive field types. After successfully encoding, we check that we are able
+// to decode the output and arrive at the same fields.
```

Comment:
Could you explain why it is better to test like this rather than in isolation?

Reply:
the primary reason is just that there's a lot of boiler plate that can be reused. another reason is because the `buf` is reused across all encoding/decodings, which helps detect if there are any side effects of fields on each other.

mainly the point of this pr is just to verify that the encodings work properly, tho i agree individual tests be a useful alternative

Reply:
Ok. A comment about this would be nice in the code, but not strictly necessary.